radicle-seed/src/lib.rs

```diff
@@ -0,0 +1,189 @@
+#[macro_use]
+extern crate async_trait;
+
+use std::{net::SocketAddr, vec};
+use thiserror::Error;
+
+use radicle_keystore::sign::ed25519;
+
+use librad::{
+    git,
+    keys,
+    meta::project,
+    net::{
+        discovery,
+        gossip,
+        gossip::types::PeerInfo,
+        peer::{self, PeerApi, PeerConfig},
+        protocol::ProtocolEvent,
+    },
+    paths,
+    peer::PeerId,
+    uri::RadUrn,
+};
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    Storage(#[from] git::storage::Error),
+
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+
+    #[error(transparent)]
+    Bootstrap(#[from] peer::BootstrapError),
+
+    #[error(transparent)]
+    Accept(#[from] peer::AcceptError),
+}
+
+#[derive(Clone)]
+pub struct Signer {
+    key: keys::SecretKey,
+}
+
+impl Signer {
+    fn new() -> Self {
+        let key = keys::SecretKey::new();
```

Comment:
> Also, Signer should actually be implemented for SecretKey in librad (ie. move it there if it isn't already).

Ya it's implemented there :)

Reply:
Done in 9d04d7a

However, trying to test this with a secret key generated from `SecretKey` gives me:
```
thread 'main' panicked at 'A valid PKCS#8 document is valid. qed: CouldNotParseKeyPair', /home/cloudhead/src/radicle-dev/radicle-link/librad/src/net/tls.rs:65:28
```

It does parse properly though:
```
> echo -n 'cdb47a4eefeb8922d953e312785521e97f5bc0b9ce66ba07037c0f1086f9a37d' | cargo run -- --root .
```

Reply:
:thinking: What's that string you're `echo`ing? The `Display` impl for `SecretKey` outputs the public part (for whatever reason, and hex-encoded for wtf reason), which looks suspiciously like this. What `SecretKey::from_bytes_and_meta` expects are the actual (RFC 8032) bytes (I know `SecStr` is an unfortunate name, it's actually `type SecStr = SecVec<u8>`).

It's been bugging me for a while that there doesn't seem to be a way to validate that a byte slice is an actual ed25519 key (public or secret) -- libraries just check if it's 32 bytes. Perhaps the proper way is to pass the public key as a "checksum" as a CLI option.

Reply:
Aha! Yeah it's basically `println!("{}", SecretKey::new())`, so that explains it. I guess it's a valid curve point but not a valid private key, or something like that. The checksum idea is interesting. Otherwise it could work a bit like `ssh -i`, where it would look for the `.pub`..

Reply:
> I guess it's a valid curve point

It's literally only checked if the input is 64 bytes long. `ed25519-zebra` uses the seed value for byte representations, which is safer, but not usable for us as we need to be able to convert to other wrapper types.


> Otherwise it could work a bit like `ssh -i`

There is actually another subtlety: once the `Storage` has been initialised, it is [tied to the key used for that](https://github.com/radicle-dev/radicle-link/blob/895e73438e6ec910d2054ad80e21571305b2c716/librad/src/git/storage.rs#L501). Perhaps we should [bootstrap](https://github.com/radicle-dev/radicle-link/blob/895e73438e6ec910d2054ad80e21571305b2c716/librad/src/net/peer.rs#L304-L313) the storage first, as the error might be more meaningful.

So yeah, I'm not sure. I guess the best we can do when constructing a `SecretKey` from a slice is to sign some random data, and see if that verifies.

Reply:
> the seed value

Well, disregard this, it's the same thing: 32 random bytes